packages/lit-element/src/lib/platform-support.ts

```diff
@@ -0,0 +1,388 @@
+            v instanceof CSSStyleSheet
+              ? Array.from(v.cssRules).reduce(
+                  (a: string, r: CSSRule) => (a += r.cssText),
+                  ''
+                )
+              : v.cssText
+          );
+          window.ShadyCSS?.ScopingShim?.prepareAdoptedCssText(css, name);
+        }
+      }
+    },
+    /**
+     * Patch connectedCallback to apply ShadyCSS custom properties shimming.
+     */
+    __baseConnectedCallback: LitElement.prototype.connectedCallback,
+    connectedCallback(this: LitElementStandIn) {
+      this.__baseConnectedCallback();
+      // Note, must do first update separately so that we're ensured
+      // that rendering has completed before calling this.
+      if (this.hasUpdated) {
+        window.ShadyCSS!.styleElement(this);
+      }
+    },
+
+    /**
+     * Patch update to apply ShadyCSS custom properties shimming for first
+     * update.
+     */
+    __baseUpdate: LitElement.prototype.update,
+    update(this: LitElementStandIn, changedProperties: unknown) {
+      const isFirstUpdate = !this.hasUpdated;
+      this.__baseUpdate(changedProperties);
+      // Note, must do first update here so rendering has completed before
+      // calling this and styles are correct by updated/firstUpdated.
+      if (isFirstUpdate) {
+        window.ShadyCSS!.styleElement(this);
+      }
+    },
+  });
+};
+
+interface ShadyTemplateResult {
+  __renderIntoShadowRoot?: boolean;
+  _$litType$?: string;
+}
+
+interface NodePartStandIn {
+  new (...args: any[]): NodePartStandIn;
+  _value: unknown;
+  _startNode: ChildNode;
+  _endNode: ChildNode | null;
+  options: RenderOptions;
+  _setValue(value: unknown): void;
+  __baseSetValue(value: unknown): void;
+}
+
+interface TemplateStandIn {
+  new (...args: any[]): TemplateStandIn;
+  _createElement(html: string): HTMLTemplateElement;
+}
+
+/**
+ * lit-html patches. These properties cannot be renamed.
+ * * NodePart.prototype._getTemplate
+ * * NodePart.prototype._setValue
+ */
+(globalThis as any)['litHtmlPlatformSupport'] = ({
+  NodePart,
+  Template,
+}: {
+  NodePart: NodePartStandIn;
+  Template: TemplateStandIn;
+}) => {
+  if (!needsPlatformSupport) {
+    return;
+  }
+
+  // console.log(
+  //   '%c Making lit-html compatible with ShadyDOM/CSS.',
+  //   'color: lightgreen; font-style: italic'
+  // );
+
+  const styledScopes: Set<string> = new Set();
+  const scopeCssStore: Map<string, string[]> = new Map();
+  let currentScope: string;
+
+  const isScopeStyled = (name: string | undefined) =>
+    needsPlatformSupport && name !== undefined ? styledScopes.has(name) : true;
+
+  const cssForScope = (name: string) => {
+    if (isScopeStyled(name)) {
+      return;
+    }
+    let scopeCss = scopeCssStore.get(name);
+    if (scopeCss === undefined) {
+      scopeCssStore.set(name, (scopeCss = []));
+    }
+    return scopeCss;
+  };
+
+  const ensureStyles = (name: string, template: HTMLTemplateElement) => {
+    if (isScopeStyled(name)) {
```

Comment:
The logic/relationship between `ensureStyles` and `styledScopes` and `isScopeStyled` is pretty circular and hard to follow. I might need an explanation.

Reply:
Changed: `ensureStyles` to `prepareStyles` and added comments.